# Lesson 6 Class Exercises: More Practice

We will use today's class exercise to reinforce the learning we've done to this point

## Get Started
Import the Numpy and Pandas packages

In [171]:
import numpy as py
import pandas as pd

## Exercise 1

Read in the apple data found in the `data/apple_data.tsv` directory.

In [172]:
apple = pd.read_csv('data/apple_data.tsv', delimiter='\t')
apple

State    1982    1987    1992    1997    2002    2007  \
0             Alabama    899     626     474     479     265     231    
1              Alaska      --      --      2       4       9      10    
2             Arizona    214     224     235     158     185     496    
3            Arkansas    512     451     247     238     164     153    
4          California  2,904   2,787   3,082   3,454   2,120   2,074    
5            Colorado    608     595     542     493     361     468    
6         Connecticut    309     272     295     242     247     248    
7            Delaware     24      18      16      18      17      23    
8             Florida    196     149      92      81      53     126    
9             Georgia    893     669     525     428     305     286    
10             Hawaii      --      --      --      --      2       9    
11              Idaho    411     377     374     342     315     321    
12           Illinois    902     824     756     651     501     444    
13            Indiana    792     705     685     615     442     265    
14               Iowa    439     458     450     443     344     364    
15             Kansas    416     318     264     261     192     147    
16           Kentucky  1,022     916     878     710     616     623    
17          Louisiana    122     142      81      94      51      35    
18              Maine    400     386     388     349     374     345    
19           Maryland    551     453     392     338     288     252    
20      Massachusetts    452     526     477     399     338     369    
21           Michigan  2,788   2,516   2,473   2,290   1,750   1,772    
22          Minnesota    391     411     476     536     488     569    
23        Mississippi    413     293     258     248     159      80    
24           Missouri    871     673     539     524     485     438    
25            Montana    142     145     160     148     187     186    
26           Nebraska    100     124     126     106     107     125    
27             Nevada     50      56      45      47      45      35    
28      New Hampshire    219     209      227    210     183     215    
29         New Jersey    509     563     522     485     525     471    
30         New Mexico    740     666     748     806     907     807    
31           New York  2,055   1,726   1,578   1,557   1,447   1,350    
32     North Carolina  1,500   1,186     990     949     598     673    
33       North Dakota     38      29      32      28       5      21    
34               Ohio  1,857   1,614   1,454   1,404   1,282   1,000    
35           Oklahoma    695     582     392     427     239     178    
36             Oregon  2,053   1,784   1,684   1,763   1,485   1,330    
37       Pennsylvania  2,580   2,270   1,854   2,183   1,777   1,886    
38       Rhode Island     68      72      61      51      58      73    
39     South Carolina    390     321     330     292     339     205    
40       South Dakota     48      60      39      40      35      60    
41          Tennessee  1,321   1,065     924   1,003     357     543    
42              Texas    845     883     782     724     596     437    
43               Utah    635     568     536     485     388     457    
44            Vermont    243     217     251     237     243     264    
45           Virginia  1,400   1,160   1,057     970     746     626    
46         Washington  5,406   5,169   4,596   4,910   3,870   3,052    
47      West Virginia    712     586     505     593     463     494    
48          Wisconsin  1,027     856     965   1,006     886     935    
49            Wyoming     25      18      20      16      14      20    

        1982      1987    1992.1    1997.1    2002.1    2007.1 2002.2 2007.2  \
0     1,548     1,217     1,011       968       594       393     191    144   
1         --        --       (D)        2        13        15       8     10   
2     1,969     4,881     5,770     4,003     1,581 

Explore the data using the approach we've used thus far:
- List all of the column names
- Identify the data types
- Find the shape of the data
- Explore categorical values -- do we have the expected number of values for each?
- Get descriptive statistics for the numerical data
- Do we have missing values?
- Do we have duplicates?

In [173]:
apple.columns
#1987 has a space

Index(['       State', '1982', '1987', '1992', '1997', '2002', '2007',
       '  1982', '  1987', '1992.1', '1997.1', '2002.1', '2007.1', '2002.2',
       '2007.2', '2002.3', '2007.3'],
      dtype='object')

In [174]:
apple.dtypes
#All strings, so need to fix

       State    object
1982            object
1987            object
1992            object
1997            object
2002            object
2007            object
  1982          object
  1987          object
1992.1          object
1997.1          object
2002.1          object
2007.1          object
2002.2          object
2007.2          object
2002.3          object
2007.3          object
dtype: object

In [175]:
apple.shape

(50, 17)

In [176]:
apple['       State'].unique().shape

(50,)

In [177]:
apple.describe()

State 1982 1987  1992  1997  2002 2007   1982   1987 1992.1  \
count            50   50   50    50    50    50   50     50     50     50   
unique           50   49   48    49    49    50   49     49     49     49   
top         Alabama   --   --  392   485   265   35      --     --    (D)   
freq              1    2    2     2     2     1    2      2      2      2   

        1997.1 2002.1 2007.1 2002.2 2007.2 2002.3 2007.3  
count       50     50     50     50     50     50     50  
unique      48     48     48     46     49     46     48  
top     5,389     (D)    (D)      8    427    (D)    (D)  
freq         2      3      3      2      2      5      3

In [178]:
apple.isna()
#something is wrong

State   1982   1987   1992   1997   2002   2007    1982    1987  \
0          False  False  False  False  False  False  False   False   False   
1          False  False  False  False  False  False  False   False   False   
2          False  False  False  False  False  False  False   False   False   
3          False  False  False  False  False  False  False   False   False   
4          False  False  False  False  False  False  False   False   False   
5          False  False  False  False  False  False  False   False   False   
6          False  False  False  False  False  False  False   False   False   
7          False  False  False  False  False  False  False   False   False   
8          False  False  False  False  False  False  False   False   False   
9          False  False  False  False  False  False  False   False   False   
10         False  False  False  False  False  False  False   False   False   
11         False  False  False  False  False  False  False   False   False   
12         False  False  False  False  False  False  False   False   False   
13         False  False  False  False  False  False  False   False   False   
14         False  False  False  False  False  False  False   False   False   
15         False  False  False  False  False  False  False   False   False   
16         False  False  False  False  False  False  False   False   False   
17         False  False  False  False  False  False  False   False   False   
18         False  False  False  False  False  False  False   False   False   
19         False  False  False  False  False  False  False   False   False   
20         False  False  False  False  False  False  False   False   False   
21         False  False  False  False  False  False  False   False   False   
22         False  False  False  False  False  False  False   False   False   
23         False  False  False  False  False  False  False   False   False   
24         False  False  False  False  False  False  False   False   False   
25         False  False  False  False  False  False  False   False   False   
26         False  False  False  False  False  False  False   False   False   
27         False  False  False  False  False  False  False   False   False   
28         False  False  False  False  False  False  False   False   False   
29         False  False  False  False  False  False  False   False   False   
30         False  False  False  False  False  False  False   False   False   
31         False  False  False  False  False  False  False   False   False   
32         False  False  False  False  False  False  False   False   False   
33         False  False  False  False  False  False  False   False   False   
34         False  False  False  False  False  False  False   False   False   
35         False  False  False  False  False  False  False   False   False   
36         False  False  False  False  False  False  False   False   False   
37         False  False  False  False  False  False  False   False   False   
38         False  False  False  False  False  False  False   False   False   
39         False  False  False  False  False  False  False   False   False   
40         False  False  False  False  False  False  False   False   False   
41         False  False  False  False  False  False  False   False   False   
42         False  False  False  False  False  False  False   False   False   
43         False  False  False  False  False  False  False   False   False   
44         False  False  False  False  False  False  False   False   False   
45         False  False  False  False  False  False  False   False   False   
46         False  False  False  False  False  False  False   False   False   
47         False  False  False  False  False  False  False   False   False   
48         False  False  False  False  False  False  False   False   False   
49         False  False  False  False  False  False  False   False   False   

    1992.1  1997.1  2002.1  2007

In [179]:
apple.isna().sum()

       State    0
1982            0
1987            0
1992            0
1997            0
2002            0
2007            0
  1982          0
  1987          0
1992.1          0
1997.1          0
2002.1          0
2007.1          0
2002.2          0
2007.2          0
2002.3          0
2007.3          0
dtype: int64

In [180]:
apple.duplicated()


0     False
1     False
2     False
3     False
4     False
5     False
6     False
7     False
8     False
9     False
10    False
11    False
12    False
13    False
14    False
15    False
16    False
17    False
18    False
19    False
20    False
21    False
22    False
23    False
24    False
25    False
26    False
27    False
28    False
29    False
30    False
31    False
32    False
33    False
34    False
35    False
36    False
37    False
38    False
39    False
40    False
41    False
42    False
43    False
44    False
45    False
46    False
47    False
48    False
49    False
dtype: bool

What needs cleaning with the data?

### Clean the data

Re-read the file and make sure we have get the missing values correct.

In [181]:
apple=pd.read_csv('./data/apple_data.tsv', delimiter='\t', na_values=['--','(D)'])

In [182]:
apple.isna().sum()

       State    0
1982            2
1987            2
1992            1
1997            1
2002            0
2007            0
  1982          2
  1987          2
1992.1          3
1997.1          3
2002.1          3
2007.1          3
2002.2          0
2007.2          0
2002.3          5
2007.3          3
dtype: int64

In [183]:
#Ko's
# apple.columns=apple.keys().str.replace("\s+","", regex=True)
# apple.columns=apple.keys().str.replace("\.1","", regex=True)
# apple.columns=apple.keys().str.replace("\.2","", regex=True)
# apple.columns=apple.keys().str.replace("\.3","", regex=True)
# apple.keys()

In [184]:
apple.isna().sum()

       State    0
1982            2
1987            2
1992            1
1997            1
2002            0
2007            0
  1982          2
  1987          2
1992.1          3
1997.1          3
2002.1          3
2007.1          3
2002.2          0
2007.2          0
2002.3          5
2007.3          3
dtype: int64

In [185]:
#chris's solution
apple2 = apple.set_axis(['State', '1982_Farms', '1987_Farms', '1992_Farms', '1997_Farms', '2002_Farms', '2007_Farms', '1982_TTLAcres', '1987_TTLAcres', '1992_TTLAcres', '1997_TTLAcres', '2002_TTLAcres', '2007_TTLAcres', '2002_FarmBear_Acres', '2007_FarmBear_Acres', '2002_Bear_Acres', '2007_Bear_Acres' ], axis=1)
apple2.head()
apple2.isna().sum()

State                  0
1982_Farms             2
1987_Farms             2
1992_Farms             1
1997_Farms             1
2002_Farms             0
2007_Farms             0
1982_TTLAcres          2
1987_TTLAcres          2
1992_TTLAcres          3
1997_TTLAcres          3
2002_TTLAcres          3
2007_TTLAcres          3
2002_FarmBear_Acres    0
2007_FarmBear_Acres    0
2002_Bear_Acres        5
2007_Bear_Acres        3
dtype: int64

Make the column names correct

In [186]:
apple2.head()

State 1982_Farms 1987_Farms 1992_Farms 1997_Farms 2002_Farms  \
0        Alabama       899        626        474        479        265    
1         Alaska        NaN        NaN         2          4          9    
2        Arizona       214        224        235        158        185    
3       Arkansas       512        451        247        238        164    
4     California     2,904      2,787      3,082      3,454      2,120    

  2007_Farms 1982_TTLAcres 1987_TTLAcres 1992_TTLAcres 1997_TTLAcres  \
0       231         1,548         1,217         1,011           968    
1        10            NaN           NaN           NaN            2    
2       496         1,969         4,881         5,770         4,003    
3       153         2,276         2,120         1,430         1,048    
4     2,074        29,933        34,105        42,024        49,704    

  2002_TTLAcres 2007_TTLAcres 2002_FarmBear_Acres 2007_FarmBear_Acres  \
0          594           393                  191                 144   
1           13            15                    8                  10   
2        1,581         1,344                  118                 424   
3          658           287                  123                 116   
4       38,268        22,184                1,850               1,850   

  2002_Bear_Acres 2007_Bear_Acres  
0             386             307  
1             NaN             NaN  
2           1,459           1,249  
3             560             220  
4          35,856          20,954

Next, remove commas. We need to alter the strings of multiple columns. We know how to do this within a single column, but how do we do this with multiple columns?  Remember the `apply()` function! We can apply any function we want.  We can't use the `pd.Series.str.replace()` function though because that belongs to the Pandas Series objects, we need an independent function. We can write our own!

We can use the formal defintion (e.g.):

```python
def remoteCommas(col):
    return col.str.replace(',','')

apple_data = apple_data.apply(remoteCommas)
```

Or we can simplify it using Pythons `lambda` function definition. `lambda` is used to declare an in-line function.  It works just like a normal function but it does not have a name and is used for quick very short in-line functions.  For example:

```python
apple_data = apple_data.apply(lambda x: x.str.replace(',',''))
```

In [187]:
apple2 = apple2.apply(lambda columns: columns.str.replace(',',''))
#we had no commas in states so that was fine

In [188]:
apple2.head()

State 1982_Farms 1987_Farms 1992_Farms 1997_Farms 2002_Farms  \
0        Alabama       899        626        474        479        265    
1         Alaska        NaN        NaN         2          4          9    
2        Arizona       214        224        235        158        185    
3       Arkansas       512        451        247        238        164    
4     California      2904       2787       3082       3454       2120    

  2007_Farms 1982_TTLAcres 1987_TTLAcres 1992_TTLAcres 1997_TTLAcres  \
0       231          1548          1217          1011           968    
1        10            NaN           NaN           NaN            2    
2       496          1969          4881          5770          4003    
3       153          2276          2120          1430          1048    
4      2074         29933         34105         42024         49704    

  2002_TTLAcres 2007_TTLAcres 2002_FarmBear_Acres 2007_FarmBear_Acres  \
0          594           393                  191                 144   
1           13            15                    8                  10   
2         1581          1344                  118                 424   
3          658           287                  123                 116   
4        38268         22184                 1850                1850   

  2002_Bear_Acres 2007_Bear_Acres  
0             386             307  
1             NaN             NaN  
2            1459            1249  
3             560             220  
4           35856           20954

Remove spaces preceeding the state names

In [189]:
apple2['State'].unique()

array(['   Alabama', '   Alaska', '   Arizona', '   Arkansas',
       '   California', '   Colorado', '   Connecticut', '   Delaware',
       '   Florida', '   Georgia', '   Hawaii', '   Idaho', '   Illinois',
       '   Indiana', '   Iowa', '   Kansas', '   Kentucky',
       '   Louisiana', '   Maine', '   Maryland', '   Massachusetts',
       '   Michigan', '   Minnesota', '   Mississippi', '   Missouri',
       '   Montana', '   Nebraska', '   Nevada', '   New Hampshire',
       '   New Jersey', '   New Mexico', '   New York',
       '   North Carolina', '   North Dakota', '   Ohio', '   Oklahoma',
       '   Oregon', '   Pennsylvania', '   Rhode Island',
       '   South Carolina', '   South Dakota', '   Tennessee', '   Texas',
       '   Utah', '   Vermont', '   Virginia', '   Washington',
       '   West Virginia', '   Wisconsin', '   Wyoming'], dtype=object)

In [190]:
apple2['State'] = apple2['State'].str.replace("   ","", regex=True)

In [191]:
apple2['State'].unique()

array(['Alabama', 'Alaska', 'Arizona', 'Arkansas', 'California',
       'Colorado', 'Connecticut', 'Delaware', 'Florida', 'Georgia',
       'Hawaii', 'Idaho', 'Illinois', 'Indiana', 'Iowa', 'Kansas',
       'Kentucky', 'Louisiana', 'Maine', 'Maryland', 'Massachusetts',
       'Michigan', 'Minnesota', 'Mississippi', 'Missouri', 'Montana',
       'Nebraska', 'Nevada', 'New Hampshire', 'New Jersey', 'New Mexico',
       'New York', 'North Carolina', 'North Dakota', 'Ohio', 'Oklahoma',
       'Oregon', 'Pennsylvania', 'Rhode Island', 'South Carolina',
       'South Dakota', 'Tennessee', 'Texas', 'Utah', 'Vermont',
       'Virginia', 'Washington', 'West Virginia', 'Wisconsin', 'Wyoming'],
      dtype=object)

In [192]:
apple2['State'] = apple2['State'].str.replace('/s+','', regex=True)

In [193]:
apple2['State'].unique()

array(['Alabama', 'Alaska', 'Arizona', 'Arkansas', 'California',
       'Colorado', 'Connecticut', 'Delaware', 'Florida', 'Georgia',
       'Hawaii', 'Idaho', 'Illinois', 'Indiana', 'Iowa', 'Kansas',
       'Kentucky', 'Louisiana', 'Maine', 'Maryland', 'Massachusetts',
       'Michigan', 'Minnesota', 'Mississippi', 'Missouri', 'Montana',
       'Nebraska', 'Nevada', 'New Hampshire', 'New Jersey', 'New Mexico',
       'New York', 'North Carolina', 'North Dakota', 'Ohio', 'Oklahoma',
       'Oregon', 'Pennsylvania', 'Rhode Island', 'South Carolina',
       'South Dakota', 'Tennessee', 'Texas', 'Utah', 'Vermont',
       'Virginia', 'Washington', 'West Virginia', 'Wisconsin', 'Wyoming'],
      dtype=object)

Convert columns to numeric. Convert to floats instead of integers because NaNs are considered float values.

In [194]:
apple2.columns

Index(['State', '1982_Farms', '1987_Farms', '1992_Farms', '1997_Farms',
       '2002_Farms', '2007_Farms', '1982_TTLAcres', '1987_TTLAcres',
       '1992_TTLAcres', '1997_TTLAcres', '2002_TTLAcres', '2007_TTLAcres',
       '2002_FarmBear_Acres', '2007_FarmBear_Acres', '2002_Bear_Acres',
       '2007_Bear_Acres'],
      dtype='object')

In [195]:
apple2.iloc[:,1:] = apple2.iloc[:,1:].astype(float)

In [196]:
apple2.dtypes

State                   object
1982_Farms             float64
1987_Farms             float64
1992_Farms             float64
1997_Farms             float64
2002_Farms             float64
2007_Farms             float64
1982_TTLAcres          float64
1987_TTLAcres          float64
1992_TTLAcres          float64
1997_TTLAcres          float64
2002_TTLAcres          float64
2007_TTLAcres          float64
2002_FarmBear_Acres    float64
2007_FarmBear_Acres    float64
2002_Bear_Acres        float64
2007_Bear_Acres        float64
dtype: object

Check the types to make sure all is good

Now desribe the numeric data

In [197]:
apple2.describe()

1982_Farms   1987_Farms   1992_Farms   1997_Farms   2002_Farms  \
count    48.00000    48.000000    49.000000    49.000000    50.000000   
mean    858.06250   764.958333   691.408163   690.510204   537.060000   
std     989.94688   915.886408   851.933974   912.523975   692.712221   
min      24.00000    18.000000     2.000000     4.000000     2.000000   
25%     237.00000   222.250000   235.000000   210.000000   168.750000   
50%     531.50000   544.500000   474.000000   428.000000   338.500000   
75%     932.00000   862.750000   782.000000   724.000000   578.250000   
max    5406.00000  5169.000000  4596.000000  4910.000000  3870.000000   

        2007_Farms  1982_TTLAcres  1987_TTLAcres  1992_TTLAcres  \
count    50.000000      48.000000      48.000000      47.000000   
mean    511.820000   12302.979167   12521.354167   12416.255319   
std     606.840183   25304.382357   27199.429306   28246.151550   
min       9.000000      41.000000      33.000000      35.000000   
25%     148.500000    1383.000000    1170.000000     999.000000   
50%     333.000000    4086.500000    4536.500000    4332.000000   
75%     562.500000    8642.000000    8791.250000    7407.000000   
max    3052.000000  145630.000000  162609.000000  169107.000000   

       1997_TTLAcres  2002_TTLAcres  2007_TTLAcres  2002_FarmBear_Acres  \
count      47.000000      47.000000      47.000000            50.000000   
mean    12927.510638    9863.234043    8470.638298           421.680000   
std     34141.187355   27100.575228   25504.573629           613.202348   
min         2.000000       6.000000      15.000000             2.000000   
25%       973.500000     596.500000     373.000000           119.250000   
50%      3187.000000    2384.000000    2057.000000           232.500000   
75%      6465.500000    4202.500000    3966.000000           400.750000   
max    215463.000000  172810.000000  165215.000000          3580.000000   

       2007_FarmBear_Acres  2002_Bear_Acres  2007_Bear_Acres  
count            50.000000        45.000000        47.000000  
mean            434.320000      8917.000000      7651.595745  
std             557.364954     24400.441667     23454.296203  
min               8.000000        19.000000         3.000000  
25%             116.500000       470.000000       261.500000  
50%             264.000000      2113.000000      1769.000000  
75%             453.250000      4083.000000      3444.000000  
max            2912.000000    152241.000000    152334.000000

### Explore the Data Again

Now that we have problems with the dataframe correct, let's explore. Are our data types correct?

In [198]:
apple2.dtypes

State                   object
1982_Farms             float64
1987_Farms             float64
1992_Farms             float64
1997_Farms             float64
2002_Farms             float64
2007_Farms             float64
1982_TTLAcres          float64
1987_TTLAcres          float64
1992_TTLAcres          float64
1997_TTLAcres          float64
2002_TTLAcres          float64
2007_TTLAcres          float64
2002_FarmBear_Acres    float64
2007_FarmBear_Acres    float64
2002_Bear_Acres        float64
2007_Bear_Acres        float64
dtype: object

Now that we have numeric data we can perform a proper summary

In [199]:
apple2.describe()

1982_Farms   1987_Farms   1992_Farms   1997_Farms   2002_Farms  \
count    48.00000    48.000000    49.000000    49.000000    50.000000   
mean    858.06250   764.958333   691.408163   690.510204   537.060000   
std     989.94688   915.886408   851.933974   912.523975   692.712221   
min      24.00000    18.000000     2.000000     4.000000     2.000000   
25%     237.00000   222.250000   235.000000   210.000000   168.750000   
50%     531.50000   544.500000   474.000000   428.000000   338.500000   
75%     932.00000   862.750000   782.000000   724.000000   578.250000   
max    5406.00000  5169.000000  4596.000000  4910.000000  3870.000000   

        2007_Farms  1982_TTLAcres  1987_TTLAcres  1992_TTLAcres  \
count    50.000000      48.000000      48.000000      47.000000   
mean    511.820000   12302.979167   12521.354167   12416.255319   
std     606.840183   25304.382357   27199.429306   28246.151550   
min       9.000000      41.000000      33.000000      35.000000   
25%     148.500000    1383.000000    1170.000000     999.000000   
50%     333.000000    4086.500000    4536.500000    4332.000000   
75%     562.500000    8642.000000    8791.250000    7407.000000   
max    3052.000000  145630.000000  162609.000000  169107.000000   

       1997_TTLAcres  2002_TTLAcres  2007_TTLAcres  2002_FarmBear_Acres  \
count      47.000000      47.000000      47.000000            50.000000   
mean    12927.510638    9863.234043    8470.638298           421.680000   
std     34141.187355   27100.575228   25504.573629           613.202348   
min         2.000000       6.000000      15.000000             2.000000   
25%       973.500000     596.500000     373.000000           119.250000   
50%      3187.000000    2384.000000    2057.000000           232.500000   
75%      6465.500000    4202.500000    3966.000000           400.750000   
max    215463.000000  172810.000000  165215.000000          3580.000000   

       2007_FarmBear_Acres  2002_Bear_Acres  2007_Bear_Acres  
count            50.000000        45.000000        47.000000  
mean            434.320000      8917.000000      7651.595745  
std             557.364954     24400.441667     23454.296203  
min               8.000000        19.000000         3.000000  
25%             116.500000       470.000000       261.500000  
50%             264.000000      2113.000000      1769.000000  
75%             453.250000      4083.000000      3444.000000  
max            2912.000000    152241.000000    152334.000000

Do we have any missing values?

In [200]:
apple2.isna().sum()

State                  0
1982_Farms             2
1987_Farms             2
1992_Farms             1
1997_Farms             1
2002_Farms             0
2007_Farms             0
1982_TTLAcres          2
1987_TTLAcres          2
1992_TTLAcres          3
1997_TTLAcres          3
2002_TTLAcres          3
2007_TTLAcres          3
2002_FarmBear_Acres    0
2007_FarmBear_Acres    0
2002_Bear_Acres        5
2007_Bear_Acres        3
dtype: int64

Are missing values reasonable?  Should we remove any?

In [201]:
apple2.head()
#Don't drop rows because we haven't tidied it. 

State  1982_Farms  1987_Farms  1992_Farms  1997_Farms  2002_Farms  \
0     Alabama       899.0       626.0       474.0       479.0       265.0   
1      Alaska         NaN         NaN         2.0         4.0         9.0   
2     Arizona       214.0       224.0       235.0       158.0       185.0   
3    Arkansas       512.0       451.0       247.0       238.0       164.0   
4  California      2904.0      2787.0      3082.0      3454.0      2120.0   

   2007_Farms  1982_TTLAcres  1987_TTLAcres  1992_TTLAcres  1997_TTLAcres  \
0       231.0         1548.0         1217.0         1011.0          968.0   
1        10.0            NaN            NaN            NaN            2.0   
2       496.0         1969.0         4881.0         5770.0         4003.0   
3       153.0         2276.0         2120.0         1430.0         1048.0   
4      2074.0        29933.0        34105.0        42024.0        49704.0   

   2002_TTLAcres  2007_TTLAcres  2002_FarmBear_Acres  2007_FarmBear_Acres  \
0          594.0          393.0                191.0                144.0   
1           13.0           15.0                  8.0                 10.0   
2         1581.0         1344.0                118.0                424.0   
3          658.0          287.0                123.0                116.0   
4        38268.0        22184.0               1850.0               1850.0   

   2002_Bear_Acres  2007_Bear_Acres  
0            386.0            307.0  
1              NaN              NaN  
2           1459.0           1249.0  
3            560.0            220.0  
4          35856.0          20954.0

Do we have any duplicated data?

In [202]:
apple2.duplicated().sum()

0

Does our only categorical column have expected values?

In [203]:
apple2['State'].unique()
#or
apple2['State'].describe()

count          50
unique         50
top       Alabama
freq            1
Name: State, dtype: object

How many rows do we have per state? 

Are the number of rows per state what we expect?

### Tidy the data

What is/are the observable unit(s)?

What are the variables?

What are the observations (values)?

Next, tidy the data by melting the data. We have multiple variables in the header and we must melt them individually.  The best way to do this is to melt each variable separately, then combine them back to a single data frame. 

Do this, you will need to perform multiple melts:
1. First melt only the colums that have the number of farms and name the column `Farms`.
2. Second, melt only the columns that have the number of acres and name the column `Acres`.
3. Third, melt only the columns that have the number of bearing farms and name the column `Bearing_Farms`,
4. Fourth, melt only the columns that have the number of bearing acres and name the column `Bearing_Acres`.

You'll also want to remove the prefix (e.g. "Farms-") from each value after melting and convert the 'Year' column to an integer.

For example, after the first melt you should have a dataframe that looks like the following:

|  | State | Year | Farms |
| - | ----- | ---- | --------- |
| 0 | Alabama | 1982 | 899 |
| 1 | Alaska | 1982 | NaN |
| 2 | Arizona | 1982 | 214 |
| 3 | Arkansas | 1982 | 512 |
| 4 | California | 1982 | 2904 |

Melt #1

In [204]:
apple_melt1 = pd.melt(apple2, id_vars='State', var_name='Year_VarName',
                      value_name='Values', ignore_index=True)
apple_melt1

State     Year_VarName    Values
0          Alabama       1982_Farms     899.0
1           Alaska       1982_Farms       NaN
2          Arizona       1982_Farms     214.0
3         Arkansas       1982_Farms     512.0
4       California       1982_Farms    2904.0
..             ...              ...       ...
795       Virginia  2007_Bear_Acres   12619.0
796     Washington  2007_Bear_Acres  152334.0
797  West Virginia  2007_Bear_Acres    4424.0
798      Wisconsin  2007_Bear_Acres    5136.0
799        Wyoming  2007_Bear_Acres      44.0

[800 rows x 3 columns]

Melt #2

In [208]:
apple_melt2 = apple_melt1['Year_VarName'].str.split('_', n=1, expand = True)
apple_melt3 = pd.concat([apple_melt1, apple_melt2], axis=1)
apple_melt3.columns = ['State', 'Year_VarName', 'Value', 'Year', 'VarName']
apple_melt3.head()

State Year_VarName   Value  Year VarName
0     Alabama   1982_Farms   899.0  1982   Farms
1      Alaska   1982_Farms     NaN  1982   Farms
2     Arizona   1982_Farms   214.0  1982   Farms
3    Arkansas   1982_Farms   512.0  1982   Farms
4  California   1982_Farms  2904.0  1982   Farms

Melt #3

Melt #4

Now combine all four melted dataframes into a single dataframe. To do this we can use the [pd.DataFrame.join()](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.join.html) function. We will need to join one table at a time.  

Note: when you join tables they must have the same indexes. So, be sure to set the index of each dataframe before joining by using the [set_index()](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.set_index.html) function.

In [ ]:
# Set the indexes to be the same


In [ ]:
# Join all of the tables.


What's the shape of the data now?

Do we have any duplicated rows now?

Now, do we want to remove missing values?

### Explore the Data Again

Now that we have our data fully cleaned and tidy'ed let's explore it to make sure all looks good. First, we should still have 50 states. Can you find out?

How many years do we have per state? We should have the same number except for the states where we removed rows with missing values

Describe the data, but make sure it does so by state.